api/client/samples/similar_regions/similar_region.py

```diff
@@ -150,11 +150,11 @@ def dist_function(x,y):
         self._logger.info("Constructing BallTrees...")
         self.metric_object = DistanceMetric.get_metric('pyfunc', **{'func':dist_function})
         self.ball = BallTree(self.data, metric=self.metric_object, leaf_size=2)
+        # TODO: don't reset the following dicts, keep ball trees and regions keyed by (search_region, level) instead of
+        # only level for more reuse
         self.balls_on_level = {}
         self.regions_on_level = {}
-        # separate ball trees to process requests on givel level
-        # shouldn't generally make separate data copies since region_id's are organized with countries < provinces < districts
-        # so data for each level should be continuous arrays
+        self.already_built = set()
```

Comment:
We rebuild if search region OR requested_level are not present in already_bult. If it is search_region which changed, we should clear everything (all trees on all levels become invalid). If it is only requested_level, we can afford to keep other trees around. With this change we clear all trees regardless. This might be less efficient but should be safe.

Reply:
yeah, the point is  I think we can avoid rebuilding if search region is changed but for now that is the case. see TODO comment above.